### PR TITLE
New: `no-restricted-syntax` rule to forbid certain syntax

### LINKED ISSUE #2422: no-try-catch rule?
try-catch statements can be problematic in async code on the server. 

It's generally best for developers to identify any source of thrown errors in libraries they consume that should be caught and handled and make a small wrapper that makes it safe to use that library everywhere without worrying about throwing. This could possibly include converting a thrown error into a NodeJS style callback. 

I propose a rule that is off-by-default (maybe on by default for nodejs env?) that warns when you use try-catch, leading you to explicitly disable the rule with a directive when you deliberately want to use try-catch.

A PR for this already exists here https://github.com/eslint/eslint/pull/2421 but we can discuss the rule in this issue.

Thoughts?

### PR DESCRIPTION
**Pull request by BYK**
Fixes #2422

### CODE REVIEW

File: docs/rules/no-restricted-syntax.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Disallow certain syntax (no-restricted-syntax)
+
+Some code bases prefer to not use certain features of JavaScript such as `FunctionExpression` or `WithStatement`. This rule warns on the passed node types.
```
Comment (nzakas) [inline docs/rules/no-restricted-syntax.md]: Need more here. This needs to talk about ASTs and ping to the parser demo to help people figure out AST node types.

File: lib/rules/no-restricted-syntax.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * @author Burak Yigit Kaya
+ * @copyright 2015 Burak Yigit Kaya. All rights reserved.
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+var nodeTypes = require("espree/lib/ast-node-types");
```
Comment (nzakas) [inline lib/rules/no-restricted-syntax.js]: Use `espree.Syntax` instead.

File: docs/rules/no-restricted-syntax.md (hunk 1 above)
Comment (BYK) [inline docs/rules/no-restricted-syntax.md]: Okay, I'll try harder. That said if you have even a rough version in your mind, I'd welcome the help :)

File: lib/rules/no-restricted-syntax.js (hunk 2 above)
Comment (BYK) [inline lib/rules/no-restricted-syntax.js]: Done.

File: docs/rules/no-restricted-syntax.md (hunk 1 above)
Comment (nzakas) [inline docs/rules/no-restricted-syntax.md]: Everything wants me to write their docs :) 

Just think about it from the point of view of someone who is new to ESLint - what would they need to know to be able to use this rule? That's what should be here.

Comment (BYK) [inline docs/rules/no-restricted-syntax.md]: Haha, not really wanting you to write docs for me :) Sorry if this implied that. I just asked if you had a specific direction in your head.

I'll make this better ;)

Comment (BYK) [inline docs/rules/no-restricted-syntax.md]: Did some more writing. Hope the new one is more informative.

File: lib/rules/no-restricted-syntax.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+};
+
+module.exports.schema = {
+    "type": "array",
+    "items": [
+        {
+            "enum": [0, 1, 2]
+        },
+        {
+            "enum": Object.keys(nodeTypes).map(function(k) {
```
Comment (BYK) [inline lib/rules/no-restricted-syntax.js]: Realized that this would prevent any new node types introduced/used by other parsers from being set. Thoughts?

Comment (nzakas) [inline lib/rules/no-restricted-syntax.js]: Good point. Let's start with this and see what feedback we get. It's always easier to start stricter and then loosen than it is to start looser and get stricter.

**Pull request merged**